cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -256,23 +256,24 @@ static ICACHE_RAM_ATTR void timer1Interrupt() {
         // Check for toggles
         int32_t cyclesToGo = wave->nextServiceCycle - now;
         if (cyclesToGo < 0) {
+          cyclesToGo = -((-cyclesToGo) % (wave->nextTimeHighCycles + wave->nextTimeLowCycles));
           waveformState ^= mask;
           if (waveformState & mask) {
             if (i == 16) {
               GP16O |= 1; // GPIO16 write slow as it's RMW
             } else {
               SetGPIO(mask);
             }
-            wave->nextServiceCycle = now + wave->nextTimeHighCycles;
-            nextEventCycles = min_u32(nextEventCycles, wave->nextTimeHighCycles);
+            wave->nextServiceCycle = now + wave->nextTimeHighCycles + cyclesToGo;
+            nextEventCycles = min_u32(nextEventCycles, min_u32(wave->nextTimeHighCycles + cyclesToGo, 1));
```

Comment:
What this and the other branch seem to be doing is to be preserving the waveform phase at expense of period jitter.

Could it be that the pulsing you're seeing on some LED boards is due to periods of some pins slightly increasing (when the total on-time is so small that a few 10s of 80mhz instructions would actually increase the %on-time appreciably)?  And this, by introducing a feedback on the period makes things average out by continually adding jitter to the last-checked pin so that over large amounts of time the `cyclesToGo` overrun is subtracted off of the hi/low time?

I think it's a fair thing to do, given that when cyclesToGo < 0 then bad stuff has already happened, actually.  Feels like a good idea.

Reply:
Indeed, if `cyclesToGo` is already significantly negative, then we already missed the right timing for the coming edge. Either we keep timing of the next edge, at the expense of phase shifting (original code), or we compensate the missed timing and shorten the next portion to keep phase stable (new version). 

Missing one cycle is not visible, whereas shifting the phase creates a visible artefact.

Also `min_u32(wave->nextTimeHighCycles + cyclesToGo, 1)` is to make sure that we trigger the change at the next microsecond. My first version without the `min_u32` resulted in missing an edge at some rare occasion and creating flashes. It looks safe with this now.

Reply:
Sorry to interfere, but from my work on the concurring PR #7022 I believe recomputing the nextServiceCycle based on `now`, instead of incrementally, is at the heart of the problem. I am also not even giving the Timer1 a head start anymore, because any delay until a specific pin is reached should in general remain the same for each transition.
So, in the context of this here PR, what happens if you don't do the cyclesToGo compensation you are implementing, but instead of
`wave->nextServiceCycle = now + wave->nextTimeHighCycles;`
use
`wave->nextServiceCycle += wave->nextTimeHighCycles;`
? (Same for low cycle)

Reply:
Thanks. I will try your PR #7022 not before 2-3 days.
@dok-net IRAM is a very precious resource, I realized your patch is actually smaller in IRAM that the current code with my patch. So that's a super good news.

Reply:
@s-hadinger I'm sorry, though mitigated by fix and feature, the IRAM size is actually larger in my PR, from master
```
IROM   *: 228428          - code in flash         (default or ICACHE_FLASH_ATTR)
IRAM   *: 27400   \ 32768 - code in IRAM          (ICACHE_RAM_ATTR, ISRs...)
DATA   *: 1252  )         - initialized variables (global, static) in RAM\HEAP
RODATA *: 680   ) \ 81920 - constants             (global, static) in RAM\HEAP
BSS    *: 25168 )         - zeroed variables      (global, static) in RAM\HEAP
```
to PR #7022
```
IROM   *: 228460          - code in flash         (default or ICACHE_FLASH_ATTR)
IRAM   *: 27492   \ 32768 - code in IRAM          (ICACHE_RAM_ATTR, ISRs...)
DATA   *: 1252  )         - initialized variables (global, static) in RAM\HEAP
RODATA *: 680   ) \ 81920 - constants             (global, static) in RAM\HEAP
BSS    *: 25240 )         - zeroed variables      (global, static) in RAM\HEAP
```
A prominent fix is that 100% duty or off cycle now work without pulsating on the output.
A specified duration now also leaves the output at the state it is at that moment, so no more force to off when stopped during duty cycle. One can always force thereafter via `digitalWrite(pin, <state>)`, after all. One might call this a breaking change, I consider it a bug fix.
Harder to fix without further increasing RAM usage is maintaining phase when duty cycle changes, but period not: (duty_before+off_before) == (duty_new + off_new). Is this something you consider a requirement?

Reply:
Thanks, I got wrong IRAM calculation. Tasmota does not use time-limited waveform, so no breaking change for us here.
Change of phase when duty cycle changes is not a major issue for LEDs, only phase change at constant duty cycle is. Having subtle flickering during fading will be hard to notice and only transient.

Reply:
I tested PR #7022 with Tasmota LEDs, I confirm there is no flickering with your PR either.

Reply:
@dok-net Actually the phase change when changing duty cycle is visible. When fading from bright to low-brightness, the fading seems to bounce on the low value and increase a bit when stabilizing. I don't see this artefact with my proposed simple patch.

@earlephilhower Would it be possible to merge this PR first, as it should have very limited impact, and consider more deeply #7022 with the time needed to see its impacts?

Reply:
@s-hadinger I think this PR should not be merged, on the grounds that
```
min_u32(nextEventCycles, min_u32(wave->nextTimeHighCycles + cyclesToGo, 1));
```
is near 100% of the time 1, but otherwise always 0.
A quick inspection seems to indicate that this causes another forced loop-iteration, executing
```
       } else {
          uint32_t deltaCycles = wave->nextServiceCycle - now;
          nextEventCycles = min_u32(nextEventCycles, deltaCycles);
        }
```
which more or less revokes the effect of the change in this PR a few CPU cycles later.
So whatever is observed as an improvement is a side-effect of iterating inside the NMI handler.
Am I seriously missing something here *headscratch*?

Reply:
> A prominent fix is that 100% duty or off cycle now work without pulsating on the output.

Err, how would a 100% duty cycle actually get into this section of code, though?  The Tone is by definition a 50% cycle and analogWrite checks for 100% duty cycle and just does `digitalWrite`s to avoid any CPU use at all.

> A specified duration now also leaves the output at the state it is at that moment, so no more force to off when stopped during duty cycle. One can always force thereafter via `digitalWrite(pin, <state>)`, after all. One might call this a breaking change, I consider it a bug fix.

That would be a breaking change.  The original code special-cased it and there are probably HW installations out there depending on this behavior.  A 3.0 tweak, not a 2.x one.

> Harder to fix without further increasing RAM usage is maintaining phase when duty cycle changes, but period not: (duty_before+off_before) == (duty_new + off_new). Is this something you consider a requirement?

The API specifies a frequency and has no guarantees about phase relationships.  You can't know what phases there are, in general, since the app code is running async to the timer...

> min_u32(nextEventCycles, min_u32(wave->nextTimeHighCycles + cyclesToGo, 1));

As @dok-net says, this does always equal 0 or 1.  I think you meant `1` to be `microsecondsToClockCycles(1)`.  However, that still means it hits the NMI at least every 1us (1 million times a second!) and not only on edges which will eat CPU like mad.

Reply:
What I know is that `min_u32(nextEventCycles, wave->nextTimeHighCycles + cyclesToGo);` does not work, because can introduce a negative number causing an edge transition to be skipped. So there's maybe a more elegant way to force it to be non-negative.

Reply:
@s-hadinger a much reworked version is in #7022 now. Changing duty cycle maintains phase if period duration remains the same. Runtime is honored precisely, not only on next incidental timer event. Overall stability, correctness and performance fixes, "works for me" (buzzer, servo, PWM PC fan control). I think that's good for a start.
I don't have measuring equipment to graph timings etc, so basically, my observations that the PR is a major improvement for my use cases is all I can provide in the way of proofing.

Reply:
@earlephilhower 

> Err, how would a 100% duty cycle actually get into this section of code, though?

First, in the case of direct use of waveform :-) Second, loss of phase for `analogWrite()`ing min or max values seemed unacceptable to me (magnitude: bug), this here PR seems to confirm that sentiment. I think an explicit from-sketch `digitalWrite()` to fully stop the PWM generator is correct use.

> That would be a breaking change. The original code special-cased it and there are probably HW installations out there depending on this behavior.

Hem. `stopWaveform()` in master just removes the pin from signal generator, effectively, asynchronously, leaving the pin in a random state, right? OTOH, `startWaveform()` to set a finish time to an already running waveform cancels the generation at an unpredictable time in master, but is well defined in PR #7022 - at least per my objective. In master, this has nasty effects like in PR #7084 whenever the duty cycle is cut short, for instance. If the expectation is  that `startWaveform()` with a running time is always used on first call for a given inactive pin, and the running time is a multiple of period, then there is no functional change between master and #7022.
Is there a need to stay compatible to broken uses?